librz/util/ht/ht_inc.c

```diff
@@ -512,3 +513,196 @@ RZ_API void Ht_(foreach)(RZ_NONNULL HtName_(Ht) *ht, RZ_NONNULL HT_(ForeachCallb
+	if (state->ti >= state->ht->size) {
+		// Iteration is done. No elements left to select.
+		return NULL;
+	}
+	// Iterate over tables until a table with an element is found.
+	for (; state->ti < state->ht->size; state->ti++) {
+		if (state->ht->table[state->ti].count == 0) {
+			// Table has no elements. Check next table.
+			continue;
+		}
+		if (state->bi < state->ht->table[state->ti].count) {
+			// Table has elements, select the element.
+			state->kv = &state->ht->table[state->ti].arr[state->bi];
+			// For the next iteration, increment bucket index to the following element.
+			state->bi++;
+			return (const VALUE_TYPE *)&state->kv->value;
+		}
+		// Reset bucket index to first bucket.
+		state->bi = 0;
+		// Go to next table
+	}
+	// Iteration is done. No elements left to select.
+	return NULL;
+}
+
+/**
+ * \brief Advances an RzIterator over a hash table to the next key in
+ * the hash table returns it as const pointer.
+ *
+ * \param it The next key as immutable or NULL if iteration terminated.
+ */
+RZ_API const KEY_TYPE *Ht_(iter_next_key)(RzIterator *it) {
+	rz_return_val_if_fail(it, NULL);
+
+	HT_(IterState) *state = it->u;
+	if (state->ti >= state->ht->size) {
+		// Iteration is done. No elements left to select.
+		return NULL;
+	}
+	// Iterate over tables until a table with an element is found.
+	for (; state->ti < state->ht->size; state->ti++) {
+		if (state->ht->table[state->ti].count == 0) {
+			// Table has no elements. Check next table.
+			continue;
+		}
+		if (state->bi < state->ht->table[state->ti].count) {
+			// Table has elements, select the element.
+			state->kv = &state->ht->table[state->ti].arr[state->bi];
+			// For the next iteration, increment bucket index to the following element.
+			state->bi++;
+			return (const KEY_TYPE *)&state->kv->key;
+		}
+		// Reset bucket index to first bucket.
+		state->bi = 0;
+		// Go to next table
+	}
+	// Iteration is done. No elements left to select.
+	return NULL;
+}
+
+RZ_API HT_(IterMutState) *Ht_(new_iter_mut_state)(RZ_NONNULL HtName_(Ht) *ht) {
+	rz_return_val_if_fail(ht, NULL);
+	HT_(IterMutState) *state = RZ_NEW0(HT_(IterMutState));
+	rz_return_val_if_fail(state, NULL);
+	state->ht = ht;
+	return state;
+}
+
+RZ_API HT_(IterState) *Ht_(new_iter_state)(const RZ_NONNULL HtName_(Ht) *ht) {
+	rz_return_val_if_fail(ht, NULL);
+	HT_(IterState) *state = RZ_NEW0(HT_(IterState));
+	rz_return_val_if_fail(state, NULL);
+	state->ht = ht;
+	return state;
+}
+
+RZ_API void Ht_(free_iter_mut_state)(HT_(IterMutState) *state) {
+	if (state) {
+		free(state);
+	}
+}
+
+RZ_API void Ht_(free_iter_state)(HT_(IterState) *state) {
+	if (state) {
+		free(state);
+	}
+}
+
+/**
+ * \brief Returns an iterator over the hash table \p ht. The iterator yields mutable values.
+ *
+ * \param ht The hash table to create the iterator for.
+ *
+ * \return The iterator over the hash table values or NULL in case of failure.
+ */
+RZ_API RZ_OWN RzIterator *Ht_(as_iter_mut)(RZ_NONNULL HtName_(Ht) *ht) {
```

Comment:
Wouldn't `RzIterator<HtName_(Ht)>` make more sense, since `Iterator` are generic containers in my understanding